### PR TITLE
[SPARK-8920] [MLlib] Add @since tags to mllib.linalg

### PR DESCRIPTION
**Pull request by sabhyankar**

### CODE REVIEW

**Comment by mengxr** [comment]
ok to test

**Comment by mengxr** [comment]
@MechCoder Could you help review this PR? Thanks!

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 1]
```diff
@@ -30,6 +30,7 @@ import org.apache.spark.sql.types._
 
 /**
  * Trait for a local matrix.
+ * @since 1.2.0
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Are you sure that a sealed trait should be tagged?

**Comment by sabhyankar** [comment]
@MechCoder Thanks for pointing that out. I agree that the tag might not be necessary for the sealed trait. Let me know if you come across anything else.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 2]
```diff
@@ -416,6 +438,7 @@ object DenseMatrix {
    * @param numCols number of columns of the matrix
    * @param rng a random number generator
    * @return `DenseMatrix` with size `numRows` x `numCols` and values in U(0, 1)
+   * @since 1.2.0
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: It seems that these methods were added to the `DenseMatrix` object in 1.3.0 right?

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 3]
```diff
@@ -513,6 +539,7 @@ class SparseMatrix(
    * @param rowIndices the row index of the entry. They must be in strictly increasing
    *                   order for each column
    * @param values non-zero matrix entries in column major
+   * @since 1.3.0
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Is tagging the version for the alternative constructor useful for the end user? If it is, then it should be done for all uniformly.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 4]
```diff
@@ -603,6 +639,7 @@ class SparseMatrix(
   /**
    * Generate a `DenseMatrix` from the given `SparseMatrix`. The new matrix will have isTransposed
    * set to false.
+   * @since 1.3.0
    */
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: The tags for numNonzeros and numActives are missing.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 5]
```diff
@@ -839,6 +885,7 @@ object Matrices {
    * @param numRows number of rows of the matrix
    * @param numCols number of columns of the matrix
    * @return `Matrix` with size `numRows` x `numCols` and values of zeros
+   * @since 1.3.0
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: I think you have interchanged the tags for `DenseMatrix` and `Matrix`

Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: For this and all other methods below.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 6]
```diff
@@ -359,6 +367,7 @@ object Vectors {
    * @param vector input vector.
    * @param p norm.
    * @return norm in L^p^ space.
+   * @since 1.2.0
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: This was made public only in version  1.3.0

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 7]
```diff
@@ -524,6 +534,7 @@ object Vectors {
 
 /**
  * A dense vector represented by a value array.
+ * @since 1.0.0
  */
 @SQLUserDefinedType(udt = classOf[VectorUDT])
 class DenseVector(val values: Array[Double]) extends Vector {
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: I think you can add tags to methods like size, toString etc as well

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 8]
```diff
@@ -571,6 +585,9 @@ class DenseVector(val values: Array[Double]) extends Vector {
 
   override def numActives: Int = size
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: tag missing

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 9]
```diff
@@ -597,6 +617,9 @@ class DenseVector(val values: Array[Double]) extends Vector {
     new SparseVector(size, ii, vv)
   }
 
+  /**
+   * @since 1.4.0
+   */
   override def argmax: Int = {
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: This was made public only in 1.5.0

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 10]
```diff
@@ -723,6 +771,9 @@ class SparseVector(
     }
   }
 
+  /**
+   * @since 1.4.0
+   */
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: This also was added recently

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/DistributedMatrix.scala [hunk 11]
```diff
@@ -21,6 +21,7 @@ import breeze.linalg.{DenseMatrix => BDM}
 
 /**
  * Represents a distributively stored matrix backed by one or more RDDs.
+ * @since 1.0.0
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/DistributedMatrix.scala]: not sure if this is to be added as well.

**Comment by MechCoder** [comment]
Thanks for the PR. I'm also not sure if tags should be added for hashcode, equals etc since the user is not going to access these features through the documentation.

**Comment by sabhyankar** [comment]
@MechCoder thank you for the very detailed review! I have made most of the changes that you had pointed out except for the following:
- I left the tags on the alternative constructor for the public classes as I thought it would be useful to know when it was introduced.
- I left the tag in the DistributedMatrix as it is not a sealed trait and would be accessible.

In addition to the above:
- I removed the tags from the hashCode and equals methods based on your comment. The only exception is the toString method in the DenseVector and SparseVector classes.

Let me know if you spot something else. Thanks again.

**Comment by MechCoder** [comment]
> I left the tags on the alternative constructor for the public classes.

Then could you make the tags availble for all the alternative constructors. eg. DenseMatrix

**Comment by sabhyankar** [comment]
@MechCoder I couldnt find the untagged alternate constructor in DenseMatrix. Can you point me to where you see it?

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 12]
```diff
@@ -254,6 +255,7 @@ class DenseMatrix(
    * @param numRows number of rows
    * @param numCols number of columns
    * @param values matrix entries in column major
+   * @since 1.3.0
    */
   def this(numRows: Int, numCols: Int, values: Array[Double]) =
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Here?

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 13]
```diff
@@ -513,6 +539,7 @@ class SparseMatrix(
    * @param rowIndices the row index of the entry. They must be in strictly increasing
    *                   order for each column
    * @param values non-zero matrix entries in column major
+   * @since 1.3.0
    */
   def this(
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Here as well.

**Comment by MechCoder** [comment]
That's it apart from those couple of comments.

**Comment by mengxr** [comment]
We don't need to add `since` tags for methods inherited from Java `Object` like `toString`, `hashCode`, and `equals`. @sabhyankar Could you address @MechCoder 's comments? We should merge this before the 1.5 RC release.

**Comment by sabhyankar** [comment]
I am going to take a look at this and update the PR by tomorrow.

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
@MechCoder Could you make another pass? Thanks!

**Comment by sabhyankar** [comment]
I removed the tags from two of the toString method and confirmed left the tags in place for the alternate constructors as mechcoder mentioned.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala (hunk 12 above)
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: ping

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 14]
```diff
@@ -839,6 +891,7 @@ object Matrices {
    * @param numRows number of rows of the matrix
    * @param numCols number of columns of the matrix
    * @return `Matrix` with size `numRows` x `numCols` and values of zeros
+   * @since 1.3.0
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Should this be 1.2.0?

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 15]
```diff
@@ -847,6 +900,7 @@ object Matrices {
    * @param numRows number of rows of the matrix
    * @param numCols number of columns of the matrix
    * @return `Matrix` with size `numRows` x `numCols` and values of ones
+   * @since 1.3.0
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Should this be 1.2.0?

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 16]
```diff
@@ -854,6 +908,7 @@ object Matrices {
    * Generate a dense Identity Matrix in `Matrix` format.
    * @param n number of rows and columns of the matrix
    * @return `Matrix` with size `n` x `n` and values of ones on the diagonal
+   * @since 1.3.0
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Should this be 1.2.0?

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 17]
```diff
@@ -911,6 +971,7 @@ object Matrices {
    * @param vector a `Vector` that will form the values on the diagonal of the matrix
    * @return Square `Matrix` with size `values.length` x `values.length` and `values`
    *         on the diagonal
+   * @since 1.3.0
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Should this be 1.2.0?

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala (hunk 12 above)
Comment (sabhyankar) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Hi @MechCoder I sincerely apologize for my ignorance here! I am a little new to Scala and am not sure what I am missing here. I will be happy to fix it if you can provide some clarification. I have fixed the other errors that you identified below and will update the PR.

Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Oh never mind, I could notice that from the diff view :( . Looks all right to me.

Comment (sabhyankar) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: :) Thanks!

**Comment by MechCoder** [comment]
LGTM @mengxr

**Comment by mengxr** [comment]
Merged into master and branch-1.5. Thanks @sabhyankar for adding versions and @MechCoder for code review!

**Pull request closed without merging**